File: lab7/Dictionary.java
// Spencer Albrecht
// salbrech
// Lab 7
// Dictionary.java
// All of the methods needed to implement a dictionary
// with key value pairs as a binary search tree
public class Dictionary implements DictionaryInterface {

   // Private inner Node class
   private class Node {
      // Fields that store the parameters that each node will contain
      String key;
      String value;
      Node left;
      Node right;

      // Constructor for node with a key and value
      Node(String key, String value){
         this.value = value;
         this.key = key;
         this.left = null;
         this.right = null;
      }

      // Constructor without parameters
      Node() {
         this.value = null;
         this.key = null;
         this.right = null;
         this.left = null;
      }
   }

   // Fields for the Dictionary class
   private Node root;     // reference to root in the Dictionary
   private int numPairs;  // number of pairs in current Dictionary

   // Constructor for a Dictionary object. Initialized with
   // default fields
   public Dictionary() {
      root = null;
      numPairs = 0;
   }

   // Private method that is used by lookup, insert, and delete
   // to find and return a node with a specific key
   private Node findKey(Node root, String key) {
      if(root==null || key.compareTo(root.key)==0)
      return root;
      if( key.compareTo(root.key)<0 )
      return findKey(root.left, key);
      else // strcmp(k, R->key)>0
      return findKey(root.right, key);
   }

   // isEmpty()
   // pre: none
   // returns true if this Dictionary is empty, false otherwise
   public boolean isEmpty() {
      return (numPairs == 0) ? true: false;
   }

   // size()
   // pre: none
   // returns the number of entries in this Dictionary
   public int size() {
      return numPairs;
   }

   // lookup()
   // pre: none
   // returns value associated key, or null reference if no such key exists
   public String lookup(String key) {
      // Check if numItems = 0 to avoid calling findKey if not needed
      if (numPairs>0) {
         Node returnNode = findKey(root, key);
         if (returnNode!=null) {
            return returnNode.value;
         }
      }
      // Returns null if Node was not found in list or if numItems = 0
      return null;
   }

   Node findLeftmost(Node root){
      Node temp = root;
      if( temp!=null ) for( ; temp.left!=null; temp=temp.left) ;
      return root;
   }

   // findParent()
   // returns the parent of N in the subtree rooted at R, or returns NULL
   // if N is equal to R. (pre: R != NULL)
   Node findParent(Node N, Node R){
      Node P=null;
      if( N!=R ){
         P = R;
         while( P.left!=N && P.right!=N ){
            if(N.key.compareTo(P.key)<0)
            P = P.left;
            else
            P = P.right;
         }
      }
      return P;
   }

   public void printInOrder(Node root, String returnString) {
      if(root!=null){
         printInOrder(root.left, returnString);
         returnString+=root.key+" "+root.value;
         printInOrder(root.right, returnString);
      }
   }

   // insert()
   // inserts new (key,value) pair into this Dictionary
   // pre: lookup(key)==null
   public void insert(String key, String value) throws DuplicateKeyException {
      Node node, tempParent, tracer;
      if( findKey(root, key)!=null ){
         System.out.println("Dictionary Error: calling insert() on duplicate key: "+key);
      }
      node = new Node(key, value);
      tempParent = null;
      tracer = root;
      // Loop that finds the deepest parent node on the
      // correct part of the tree
      while( tracer!=null ){
         tempParent = tracer;
         if( key.compareTo(tracer.key)<0 ) {
            tracer = tracer.left;
         }
         else tracer = tracer.right;
      }
      // If the Dictionary is empty, make the new node the Root
      if(tempParent==null) {
         root = node;
      }
      //
      else if( key.compareTo(tempParent.key)<0 ) {
         tempParent.left = node;
      }
      else {
         tempParent.right = node;
      }
      numPairs++;
   }

   // delete()
   // deletes pair with the given key
   // pre: lookup(key)==null
   public void delete(String key) throws KeyNotFoundException {
      Node node, parent, temp;
      if(root==null){
         System.out.println("Dictionary Error: calling delete() on null Dictionary reference");
      }
      node = findKey(root, key);
      if( node==null ){
         System.out.println("Dictionary Error: calling delete() on null Dictionary reference");
      }
      if( node.left==null && node.right==null ){  // case 1 (no children)
         if( node==root ){
            root = null;
         }else{
            parent = findParent(node, root);
            if( parent.right==node ) parent.right = null;
            else parent.left = null;
         }
      }else if( node.right==null ){  // case 2 (left but no right child)
         if( node==root ){
            root = node.left;
         }else{
            parent = findParent(node, root);
            if( parent.right==node ) parent.right = node.left;
            else parent.left = node.left;
         }
      }else if( node.left==null ){  // case 2 (right but no left child)
         if( node==root ){
            root = node.right;
         }else{
            parent = findParent(node, root);
            if( parent.right==node ) parent.right = node.right;
            else parent.left = node.right;
         }
      }else{                     // case3: (two children: N->left!=null && N->right!=null)
         temp = findLeftmost(node.right);
         node.key = temp.key;
         node.value = temp.value;
         parent = findParent(temp, node);
         if(parent.right==temp) parent.right = temp.right;
         else parent.left = temp.right;
      }
      numPairs--;
   }

   // makeEmpty()
   // pre: none
   public void makeEmpty() {
      root = null;
      numPairs = 0;
   }

   // toString()
   // returns a String representation of this Dictionary
   // overrides Object's toString() method
   // pre: none
   public String toString() {
      String returnString = "";
      if(root!=null){
         printInOrder(root.left, returnString);
         returnString+=root.key+" "+root.value;
         printInOrder(root.right, returnString);
      }
   }

}

File: pa5/Dictionary.c
//-----------------------------------------------------------------------------
//  Spencer Albrecht
//  salbrech
//  PA 5
//  Dictionary.c
//  Implementation file for Dictionary ADT using a hash table
//-----------------------------------------------------------------------------

#include<stdio.h>
#include<stdlib.h>
#include<string.h>
#include<assert.h>
#include"Dictionary.h"
#define TABLE_SIZE 101

// rotate_left()
// rotate the bits in an unsigned int
unsigned int rotate_left(unsigned int value, int shift) {
    int sizeInBits = 8*sizeof(unsigned int);
    shift = shift & (sizeInBits - 1);
    if ( shift == 0 ) {
        return value;
    }
    return (value << shift) | (value >> (sizeInBits - shift));
}

// pre_hash()
// turn a string into an unsigned int
unsigned int pre_hash(char* input) {
    unsigned int result = 0xBAE86554;
    while (*input) {
        result ^= *input++;
        result = rotate_left(result, 5);
    }
    return result;
}

// hash()
// turns a string into an int in the range 0 to TABLE_SIZE-1
int hash(char* key){
    return pre_hash(key)%TABLE_SIZE;
}

// Creation of NodeObj data type
typedef struct NodeObj{
    char key[150];
    char value[150];
    struct NodeObj* next;
    struct NodeObj* last;
} NodeObj;

// Creation of type "Node" that points to NodeObj
typedef NodeObj* Node;

// newNode()
// constructor of the Node type
Node newNode(char* key, char* value) {
    Node node  = malloc(sizeof(NodeObj));
    if(node==NULL) {
        fprintf(stderr,
            "malloc() failed when trying to allocate for a newNode\n");
        exit(EXIT_FAILURE);
    }
    strcpy(node->key, key);
    strcpy(node->value, value);
    node->next = NULL;
    node->last = NULL;
    return(node);
}

// freeNode()
// destructor for the Node type
void freeNode(Node* pN){
    if( pN!=NULL && *pN!=NULL ){
        free(*pN);
        *pN = NULL;
    }
}

// Creation of the DictionaryObj data type
typedef struct DictionaryObj{
    int numItems;
    Node array[TABLE_SIZE];
} DictionaryObj;

// Creation of type "Dictionary" that points to DictionaryObj
typedef DictionaryObj* Dictionary;

// findKey
// returns the Node associated with the key or returns null if no
// such key k exists.
// pre: none
Node findKey(Dictionary dict, char* key) {
    // If the dict is null or is empty return null
    if (dict!=NULL && !isEmpty(dict)) {

        int index = hash(key);
        // printf("the index for key: %s is %d\n", key, index);
        // printf("line 99\n");
        // Case if there are no collisions and the node is the only
        // one at that array index
        if(dict->array[index]==NULL) {
            // printf("line 103\n");
            return NULL;
        }
        // Case if chaining is used to avoid collisions and the linked
        // list at that index needs to be searched
        else {
            Node finder = dict->array[index];
            // printf("line 110\n");
            while(finder!=NULL) {
                // Check the key on each node and compary to target.
                // Return Node object if found
                if (strcmp(finder->key, key)==0) {
                    // printf("line 115\n");
                    return finder;
                }
                else {
                    // printf("line 119\n");
                    finder = finder->next;
                }
            }
        }
    }
    // Only return null if the key was not found
    return NULL;
}

// newDictionary()
// Constructor of the Dictionary Type
Dictionary newDictionary(void) {
  Dictionary dict = malloc(sizeof(DictionaryObj));
  if(dict==NULL) {
    fprintf(stderr,
      "malloc() failed when trying to allocate for a newDictionary\n");
      exit(EXIT_FAILURE);
    }
    dict->numItems = 0;
    return dict;
}

// freeDictionary()
// destructor for the Dictionary type
void freeDictionary(Dictionary* pD){
  if( pD!=NULL && *pD!=NULL ){
    if(!isEmpty(*pD)) {
      makeEmpty(*pD);
    }
    free(*pD);
    *pD = NULL;
  }
}

// isEmpty()
// returns 1 (true) if S is empty, 0 (false) otherwise
// pre: none
int isEmpty(Dictionary D){
  if( D==NULL ){
    fprintf(stderr,
      "Dictionary Error: calling isEmpty() on NULL Dictionary reference\n");
      exit(EXIT_FAILURE);
    }
    return(D->numItems==0);
  }

// size()
// returns the number of (key, value) pairs in D
// pre: none
int size(Dictionary D){
  if( D==NULL ){
    fprintf(stderr,
      "Dictionary Error: calling size() on NULL Dictionary reference\n");
      exit(EXIT_FAILURE);
    }
    return(D->numItems);
  }

// lookup()
// returns the value v such that (k, v) is in D, or returns NULL if no
// such value v exists.
// pre: none
char* lookup(Dictionary D, char* k) {
    if( D==NULL ){
        fprintf(stderr,
            "Dictionary Error: calling lookup() on NULL Dictionary reference\n");
        exit(EXIT_FAILURE);
    }
    // Check if numItems = 0 to avoid calling findKey if not needed
    if (!isEmpty(D)) {
        // printf("line 188\n");
        Node returnNode = findKey(D,k);
        if (returnNode!=NULL) {
            // printf("line 190\n");
            return returnNode->value;
        }
    }
    // Returns null if Dictionary is empty
    // printf("line 197\n");
    return NULL;
}

// insert()
// inserts new (key,value) pair into D
// pre: lookup(D, k)==NULL
void insert(Dictionary D, char* k, char* v) {
  // printf("line 204\n");
  if( D==NULL ){
    fprintf(stderr,
      "Dictionary Error: calling insert() on NULL Dictionary reference\n");
      exit(EXIT_FAILURE);
  }
  // printf("line 209\n");
  // Check if a Node with that key is already in the list,
  // if it is, print error message and exit
  if (lookup(D,k)!=NULL) {
    fprintf(stderr,
      "Dictionary Error: calling insert() with duplicate keys\n");
      exit(EXIT_FAILURE);
  }
  else {

    int index = hash(k);

    // Case if there are no collisions and the node is the only
    // one at that array index
    if(D->array[index]==NULL) {
        Node temp = newNode(k,v);
        D->array[index] = temp;
        D->numItems++;
    }
    // Case if chaining is used to avoid collisions and the linked
    // list at that index needs to be searched
    else {
        Node temp = newNode(k,v);
        // Set the current head to the new node's next
        temp->next = D->array[index];
        // Set the current head's last to the new node
        D->array[index]->last = temp;
        D->array[index] = temp;
        D->numItems++;
    }
  }
}

// delete()
// deletes pair with the key k
// pre: lookup(D, k)!=NULL
void delete(Dictionary D, char* k){
  if( D==NULL ){
    fprintf(stderr,
      "Dictionary Error: calling delete() on NULL Dictionary reference\n");
      exit(EXIT_FAILURE);
  }
  // Prints error and exits if the key we are trying to delete is
  // not in the list
  if (lookup(D,k)==NULL) {
    fprintf(stderr,
      "Dictionary Error: calling delete() on NULL Dictionary reference\n");
      exit(EXIT_FAILURE);
  }
  else {
    // Create a new Node a set it equal to the Node we want to delete
    Node tempNode = findKey(D,k);
    int index = hash(k);
    Node tracer = D->array[index];

    // Loop that runs through link list at appropriate index
    while(tracer!=NULL) {
        if (tracer == tempNode) {
            // Case if we are deleting the first Node
            if (D->array[index]==tracer) {
                D->array[index]=NULL;
                // printf("line 269\n");
            }
            // Case if we are deleting the last Node
            else if (tracer->next==NULL) {
                tracer->last->next = NULL;
                freeNode(&tracer);
                // printf("line 277\n");
            }
            // Case if we are deleting a middle Node
            else {
                tracer->last->next = tracer->next;
                tracer->next->last = tracer->last;
                freeNode(&tracer);
                // printf("line 282\n");
            }
        }
        tracer = tracer->next;
    }
    D->numItems--;
  }
}

// makeEmpty()
// re-sets D to the empty state.
// pre: none
void makeEmpty(Dictionary D){
    if( D==NULL ){
        fprintf(stderr,
            "Dictionary Error: calling makeEmpty() on NULL Dictionary reference\n");
            exit(EXIT_FAILURE);
        }
        for (int i = 0; i < TABLE_SIZE; i++) {
            Node tracer = D->array[i];
            Node tempHead = D->array[i];
            while(tracer!=NULL) {
                tempHead = tempHead->next;
                freeNode(&tracer);
                tracer = tempHead;
            }
            D->numItems = 0;
        }
    }

// printDictionary()
// pre: none
// prints a text representation of D to the file pointed to by out
void printDictionary(FILE* out, Dictionary D){
  if( D==NULL ){
    fprintf(stderr,
      "Dictionary Error: calling printDictionary() on NULL Dictionary reference\n");
      exit(EXIT_FAILURE);
    }
    for (int i = 0; i < TABLE_SIZE; i++) {
        Node tracer = D->array[i];
        while(tracer!=NULL) {
            fprintf(out, "%s %s\n", tracer->key, tracer->value);
            tracer = tracer->next;
        }
    }
  }
